cpp/src/sort/sort.cu

```diff
@@ -17,15 +17,113 @@
+  {
+    if (col.nullable()) {
+      validity_as_column.push_back(cudf::is_valid(col));
+      validity_as_column.back()->set_null_mask(copy_bitmask(col));
+      flat_columns.push_back(validity_as_column.back()->view());
+      if (not column_order.empty()) flat_column_order.push_back(col_order);  // doesn't matter.
+      if (not null_precedence.empty()) flat_null_precedence.push_back(col_null_order);
+    }
+    for (auto it = col.child_begin(); it != col.child_end(); it++) {
+      auto const& child = *it;
+      if (child.type().id() == type_id::STRUCT) {
+        flatten_struct_column(child, col_order, null_order::BEFORE);
+        // default spark behaviour is null_order::BEFORE
+      } else {
+        flat_columns.push_back(child);
+        if (not column_order.empty()) flat_column_order.push_back(col_order);
+        if (not null_precedence.empty()) flat_null_precedence.push_back(null_order::BEFORE);
+        // default spark behaviour is null_order::BEFORE
+      }
+    }
+  }
+  // Note: possibly expand for flattening list columns too.
+
+  /**
+   * @copydoc flatten_table
+   *
+   * @return tuple with flattened table, vector of boolean columns (struct validity),
+   * flattened column order, flattened null precedence.
+   */
+  auto operator()()
+  {
+    for (decltype(input.num_columns()) i = 0; i < input.num_columns(); ++i) {
+      auto const& col = input.column(i);
+      if (col.type().id() == type_id::STRUCT) {
+        flatten_struct_column(col,
+                              (column_order.empty() ? order() : column_order[i]),
+                              (null_precedence.empty() ? null_order() : null_precedence[i]));
+      } else {
+        flat_columns.push_back(col);
+        if (not column_order.empty()) flat_column_order.push_back(column_order[i]);
+        if (not null_precedence.empty()) flat_null_precedence.push_back(null_precedence[i]);
+      }
+    }
+
+    return std::make_tuple(table_view{flat_columns},
+                           std::move(validity_as_column),
+                           std::move(flat_column_order),
+                           std::move(flat_null_precedence));
+  }
+};
+
+/**
+ * @copydoc cudf::detail::flatten_nested_columns
+ */
+std::tuple<table_view,
+           std::vector<std::unique_ptr<column>>,
+           std::vector<order>,
+           std::vector<null_order>>
+flatten_nested_columns(table_view const& input,
+                       std::vector<order> const& column_order,
+                       std::vector<null_order> const& null_precedence)
+{
+  std::vector<std::unique_ptr<column>> validity_as_column;
+  auto const has_struct = std::any_of(
+    input.begin(), input.end(), [](auto const& col) { return col.type().id() == type_id::STRUCT; });
+  if (not has_struct)
+    return std::make_tuple(input, std::move(validity_as_column), column_order, null_precedence);
+
+  return flatten_table{input, column_order, null_precedence}();
+}
```

Comment:
The flattening functionality should be moved somewhere more accessible as @hyperbolic2346 will need this for adding support to Join for struct columns. 

Maybe in one of the struct headers?

Reply:
moved to struct/utilities.hpp (it already has similar `extract_ordered_struct_children`)